cookbook/case_studies/feature_engineering/feast_integration/feast_type_transformers.py

```diff
@@ -0,0 +1,146 @@
+from feast.repo_config import RepoConfig
+from flytekit import FlyteContext
+from flytekit.core.type_engine import TypeEngine, TypeTransformer
+from flytekit.models.literals import Literal, Scalar
+from flytekit.models.types import LiteralType, SimpleType
+from feast.infra.offline_stores.file import FileOfflineStoreConfig
+from feast.infra.online_stores.sqlite import SqliteOnlineStoreConfig
+from feast import FeatureStore as FeastFeatureStore
+from google.protobuf.struct_pb2 import Struct
+from google.protobuf.json_format import MessageToDict
+from typing import Any, Dict, Iterable, List, Optional, Tuple, Union
+from feast.entity import Entity
+from feast.feature_view import FeatureView
+from feast.feature_service import FeatureService
+
+
+@dataclass_json
+@dataclass
+class FeatureStoreConfig:
+    registry_path: str
+    project: str
+    s3_bucket: str
+    online_store_path: str = 'online.db'
+
+
+@dataclass_json
+@dataclass
+class FeatureStore:
+    config: FeatureStoreConfig
+
+    def _build_feast_feature_store(self):
+        os.environ["FEAST_S3_ENDPOINT_URL"] = aws.S3_ENDPOINT.get()
+        os.environ["AWS_ACCESS_KEY_ID"] = aws.S3_ACCESS_KEY_ID.get()
+        os.environ["AWS_SECRET_ACCESS_KEY"] = aws.S3_SECRET_ACCESS_KEY.get()
+
+        config = RepoConfig(
+            registry=f"s3://{self.config.s3_bucket}/{self.config.registry_path}",
+            project=self.config.project,
+            # Notice the use of a custom provider.
+            provider="custom_provider.provider.FlyteCustomProvider",
+            offline_store=FileOfflineStoreConfig(),
+            online_store=SqliteOnlineStoreConfig(path=self.config.online_store_path),
+        )
+        return FeastFeatureStore(config=config)
+
+    def apply(
+        self,
+        objects: Union[
+            Entity,
+            FeatureView,
+            FeatureService,
+            List[Union[FeatureView, Entity, FeatureService]],
+        ],
+    ) -> None:
+        fs = self._build_feast_feature_store()
+        fs.apply(objects)
+
+        # Applying also initializes the sqlite tables in the online store
+        FlyteContext.current_context().file_access.upload(self.config.online_store_path, f"s3://{self.config.s3_bucket}/{self.config.online_store_path}")
+
+    def get_historical_features(
+        self,
+        entity_df: Union[pd.DataFrame, str],
+        features: Optional[Union[List[str], FeatureService]] = None,
+    ) -> pd.DataFrame:
+        fs = self._build_feast_feature_store()
+        retrieval_job = fs.get_historical_features(
+            entity_df=entity_df,
+            features=features,
+        )
+        return retrieval_job.to_df()
+
+    def materialize(
+        self,
+        start_date: datetime,
+        end_date: datetime,
+        feature_views: Optional[List[str]] = None,
+    ) -> None:
+        FlyteContext.current_context().file_access.download(f"s3://{self.config.s3_bucket}/{self.config.online_store_path}", self.config.online_store_path)
```

Comment:
Rather than downloading and uploading the online store, can we replace the user-given online store with an s3 bucket, just like the `registry_path`?

Reply:
We cannot due to the way the offline data is written into the sqlite tables. It's a two-step process: (1) load offline data, and then (2) write to the appropriate table. In both steps the feast sdk assumes that the files will be local, in (1) they use pyarrow to load the [offline table](https://github.com/feast-dev/feast/blob/2d52ce73309aa77ce5c76f39549023944ae3afa8/sdk/python/feast/infra/local.py#L112), and (2) uses a straighup mention to `Path` like [so](https://github.com/feast-dev/feast/blob/master/sdk/python/feast/infra/online_stores/sqlite.py#L69-L70). 

Good news is that (1) will be fixed in the next release due to https://github.com/feast-dev/feast/pull/1789/, but there's no corresponding change to the sqlite online store (i.e. issue (2) mentioned above).

Reply:
Okay